add response_hook to Cosmos methods

bryevdv opened the pull request.

adxsdk6 commented:
Can one of the admins verify this patch?

bryevdv commented:
@christopheranderson I've started with just the `CosmosClient` methods (except "get" pending the other PR). Here is sample interaction:
```
In [6]: client.delete_database(config['DATABASE'], response_hook=lambda x: print(x))
{'Cache-Control': 'no-store, no-cache', 'Pragma': 'no-cache', 'Content-Length': '0', 'Content-Type': 'application/json', 'Content-Location': 'https://brvandev-cosmos-westus2.documents.azure.com/dbs/db2/', 'Server': 'Microsoft-HTTPAPI/2.0', 'Strict-Transport-Security': 'max-age=31536000', 'x-ms-last-state-change-utc': 'Thu, 30 May 2019 18:37:57.841 GMT', 'x-ms-resource-quota': 'databases=1000;', 'x-ms-resource-usage': 'databases=2;', 'lsn': '8', 'x-ms-schemaversion': '1.8', 'x-ms-quorum-acked-lsn': '7', 'x-ms-current-write-quorum': '3', 'x-ms-current-replica-set-size': '4', 'x-ms-xp-role': '1', 'x-ms-global-Committed-lsn': '7', 'x-ms-number-of-read-regions': '0', 'x-ms-transport-request-id': '341791', 'x-ms-cosmos-llsn': '8', 'x-ms-cosmos-quorum-acked-llsn': '7', 'x-ms-session-token': '0:-1#8', 'x-ms-request-charge': '4.95', 'x-ms-serviceversion': 'version=2.2.0.0', 'x-ms-activity-id': '7e3e85ec-bcb0-4886-9e9e-c15c0a17e143', 'x-ms-gatewayversion': 'version=2.2.0.0', 'Date': 'Thu, 30 May 2019 18:46:27 GMT'}

In [7]: db = client.create_database(id=config['DATABASE'], response_hook=lambda x: print(x))
{'Cache-Control': 'no-store, no-cache', 'Pragma': 'no-cache', 'Transfer-Encoding': 'chunked', 'Content-Type': 'application/json', 'Server': 'Microsoft-HTTPAPI/2.0', 'Strict-Transport-Security': 'max-age=31536000', 'x-ms-last-state-change-utc': 'Thu, 30 May 2019 18:37:57.841 GMT', 'etag': '"0000c162-0000-0800-0000-5cf025180000"', 'x-ms-resource-quota': 'databases=1000;', 'x-ms-resource-usage': 'databases=3;', 'lsn': '11', 'x-ms-schemaversion': '1.8', 'x-ms-quorum-acked-lsn': '10', 'x-ms-current-write-quorum': '3', 'x-ms-current-replica-set-size': '4', 'x-ms-xp-role': '1', 'x-ms-global-Committed-lsn': '10', 'x-ms-number-of-read-regions': '0', 'x-ms-transport-request-id': '341809', 'x-ms-cosmos-llsn': '11', 'x-ms-cosmos-quorum-acked-llsn': '10', 'x-ms-session-token': '0:-1#11', 'x-ms-request-charge': '4.95', 'x-ms-serviceversion': 'version=2.2.0.0', 'x-ms-activity-id': '395226b4-3c80-46a3-8cff-0df68d5358f7', 'x-ms-gatewayversion': 'version=2.2.0.0', 'Date': 'Thu, 30 May 2019 18:46:48 GMT'}
```

christopheranderson commented:
Thoughts:
 * Test coverage can come from adding an additional check in the existing crud tests to make sure we're getting what we expect back
 * Passing in the headers to the Callable is probably fine, but I'd love to ship a default Callable which unpacks those headers into nice getters for things like activity_id. For example:
    ```python
    rm = ResponseMetadata()
    my_item = item.read(response_hook=rm)
    rm.activity_id
    >> "asdfasdf-a-asdf-asdf-asdf-asf"
    rm.request_charge
    >> 1.0
    rm.headers
    >> <dictionary blah blah blah>
    ```

bryevdv commented:
@christopheranderson @johanste I've added `RecordHeaders` in https://github.com/Azure/azure-sdk-for-python/pull/5557/commits/530f4016b154e46a3a4bea83f2aa737a7ebc1156 Let me know that this is what you had in mind, as well as if tests like I added the crud tests are what you had in mind.

christopheranderson commented on sdk/cosmos/azure-cosmos/azure/cosmos/diagnostics.py:
```diff
@@ -0,0 +1,86 @@
+
+from requests.structures import CaseInsensitiveDict
+
+class RecordHeaders(object):
```
I'd prefer not talking about headers with naming of the object. We will eventually want things like request diagnostics (aka Spans/metrics/etc.)

Biggest issue with this approach is different responses have have different settings. You don't get back quota/etc on readItem calls. You only get back indexing progress on container.read. You only get query metrics on queries. We either have to teach our users all our headers (which we're really not happy about; it's not very discoverable for customers and it doesn't code nicely imo) or build LOTS of these RecordHeaders things.
